File: tools/mm/pklite.py
"""PKLITE 1.15 (extra compression, large model) depacker, transliterated from the MICRO.EXE stub.

Stub layout (image offsets, image = file minus 0x60-byte MZ header):
  0x000  loader: memory check, relocate decompressor to top of memory
  0x034  XOR-chain decryptor: 0x136 words from 0x2B2 downwards, plain = c[i] ^ c[i+1], seed 0x070C
  0x058  decompressor (runs at seg:0000; tables at decompressor-relative 0x21D/0x228/0x233/0x243)
  0x2C0  compressed bitstream, then relocations, then SS SP CS IP

Bitstream: 16-bit LE words, LSB first. Flag bit 0 = literal (byte ^ remaining-bit-counter),
flag bit 1 = match. Lengths from a prefix tree with escape 0x19 (+ raw byte; 0xFE = pointer
normalisation no-op, 0xFF = end). Offsets: high byte from a prefix tree, low byte raw.
Relocations after the end marker: [count][offset*count] blocks; count 0 advances the segment by
0x0FFF paragraphs; 0xFFFF terminates. Then SS, SP, CS, IP words (segments relative to load seg).
"""
from dataclasses import dataclass, field
import struct

HEADER_LEN = 0x60
COMPRESSED_START = 0x2C0
DECRYPT_TOP = 0x2B2
DECRYPT_WORDS = 0x136
DECRYPT_SEED = 0x070C
DECOMP_BASE = 0x58


@dataclass
class Unpacked:
    image: bytes
    relocs: list                  # list of (segment, offset)
    ss: int
    sp: int
    cs: int
    ip: int
    minalloc: int = 0
    notes: dict = field(default_factory=dict)


class BitReader:
    def __init__(self, data: bytes, pos: int):
        self.data = data
        self.pos = pos
        self.buf = 0
        self.left = 0
        self._reload()

    def _reload(self):
        self.buf = self.data[self.pos] | (self.data[self.pos + 1] << 8)
        self.pos += 2
        self.left = 16

    def bit(self) -> int:
        b = self.buf & 1
        self.buf >>= 1
        self.left -= 1
        if self.left == 0:
            self._reload()
        return b

    def byte(self) -> int:
        v = self.data[self.pos]
        self.pos += 1
        return v

    def word(self) -> int:
        v = self.data[self.pos] | (self.data[self.pos + 1] << 8)
        self.pos += 2
        return v


def decrypt_stub(image: bytearray) -> None:
    prev = DECRYPT_SEED
    off = DECRYPT_TOP
    for _ in range(DECRYPT_WORDS):
        w = image[off] | (image[off + 1] << 8)
        p = w ^ prev
        image[off] = p & 0xFF
        image[off + 1] = p >> 8
        prev = w
        off -= 2


def _tables(image: bytes):
    t1 = image[DECOMP_BASE + 0x21D: DECOMP_BASE + 0x21D + 16]
    t2 = image[DECOMP_BASE + 0x228: DECOMP_BASE + 0x228 + 16]
    t3 = image[DECOMP_BASE + 0x233: DECOMP_BASE + 0x233 + 16]
    toff = image[DECOMP_BASE + 0x243: DECOMP_BASE + 0x243 + 32]
    return t1, t2, t3, toff


def _read_length(br: BitReader, t1, t2, t3):
    """Returns length, or None for end-of-data, or -1 for the 0xFE no-op."""
    bx = br.bit()
    bx = (bx << 1) | br.bit()
    if bx >= 2:
        return bx                          # codes 10, 11 -> 2, 3
    bx = (bx << 1) | br.bit()
    if bx == 0:
        return t1[0]                       # 4
    bx = (bx << 1) | br.bit()
    if bx < 5:
        return t1[bx]                      # 5, 6, 7
    bx = (bx << 1) | br.bit()
    if bx <= 0xC:
        return t1[bx]                      # 8, 9, 10
    # bx in 13..15
    bx &= 3
    bx = (bx << 1) | br.bit()
    if bx < 5:
        cl = t2[bx]                        # 11, 12, 0x19
    else:
        bx = (bx << 1) | br.bit()
        if bx <= 0xC:
            cl = t2[bx]                    # 13, 14, 15
        else:
            bx &= 3
            bx = (bx << 1) | br.bit()
            if bx < 5:
                cl = t3[bx]                # 16, 17, 18
            else:
                bx = (bx << 1) | br.bit()
                cl = t3[bx]                # 19..24
    if cl != 0x19:
        return cl
    al = br.byte()
    if al >= 0xFE:
        return None if al == 0xFF else -1
    return 0x19 + al


def _read_offset_high(br: BitReader, toff) -> int:
    if br.bit():
        return 0
    bx = 0
    for _ in range(3):
        bx = (bx << 1) | br.bit()
    if bx < 2:
        return toff[bx]
    bx = (bx << 1) | br.bit()
    if bx < 8:
        return toff[bx]
    bx = (bx << 1) | br.bit()
    if bx < 0x17:
        return toff[bx]
    bx = (bx << 1) | br.bit()
    return bx & 0xDF


def unpack(data: bytes) -> Unpacked:
    if data[:2] != b"MZ":
        raise ValueError("not an MZ executable")
    cparhdr = struct.unpack_from("<H", data, 8)[0]
    hdr = cparhdr * 16
    if hdr != HEADER_LEN or data[0x1C] != 0x0F or data[0x1D] & 0x30 != 0x30:
        raise ValueError("expected PKLITE 1.15 with extra compression + large model")
    image = bytearray(data[hdr:])
    decrypt_stub(image)
    t1, t2, t3, toff = _tables(image)

    out = bytearray()
    br = BitReader(image, COMPRESSED_START)
    while True:
        if br.bit() == 0:
            out.append(br.byte() ^ (br.left & 0xFF))
            continue
        length = _read_length(br, t1, t2, t3)
        if length is None:
            break
        if length == -1:
            continue
        if length == 2:
            off = br.byte()
        else:
            off = (_read_offset_high(br, toff) << 8) | br.byte()
        if off == 0 or off > len(out):
            raise ValueError(f"bad match offset {off} at out {len(out)}")
        start = len(out) - off
        for i in range(length):
            out.append(out[start + i])

    # relocations
    relocs = []
    seg = 0
    while True:
        count = br.word()
        if count == 0xFFFF:
            break
        if count == 0:
            seg += 0x0FFF
            continue
        for _ in range(count):
            relocs.append((seg, br.word()))
    ss = br.word(); sp = br.word(); cs = br.word(); ip = br.word()

    # memory: stub asks for 0x7DA9 paragraphs above the load segment; use that as minalloc bound
    total_paras = struct.unpack_from("<H", image, 1)[0]
    image_paras = (len(out) + 15) // 16
    # UNP 4.11 reserves 0x1D paragraphs more than the stub's request; match it so DOS memory layout
    # is identical between our output and UNP's.
    minalloc = max(0, total_paras - image_paras + 0x1D)
    return Unpacked(bytes(out), relocs, ss, sp, cs, ip, minalloc,
                    {"compressed_end": br.pos, "total_paras": total_paras})


def build_mz(u: Unpacked) -> bytes:
    """Assemble a plain MZ file: 32-byte fixed header + relocation table padded to a paragraph."""
    reloc_bytes = b"".join(struct.pack("<HH", off, seg) for seg, off in u.relocs)
    hdr_len = 0x20 + len(reloc_bytes)
    hdr_len = (hdr_len + 15) & ~15
    total = hdr_len + len(u.image)
    pages = (total + 511) // 512
    cblp = total % 512
    header = struct.pack("<2sHHHHHHHHHHHHH", b"MZ", cblp, pages, len(u.relocs), hdr_len // 16,
                         u.minalloc, 0xFFFF, u.ss, u.sp, 0, u.ip, u.cs, 0x20, 0)
    header += b"\0" * (0x20 - len(header))          # relocation table starts at e_lfarlc = 0x20
    body = header + reloc_bytes
    body += b"\0" * (hdr_len - len(body))
    return body + u.image


if __name__ == "__main__":
    import sys
    src, dst = sys.argv[1], sys.argv[2]
    u = unpack(open(src, "rb").read())
    open(dst, "wb").write(build_mz(u))
    print(f"image {len(u.image)} bytes, {len(u.relocs)} relocs, CS:IP={u.cs:04X}:{u.ip:04X} "
          f"SS:SP={u.ss:04X}:{u.sp:04X} minalloc={u.minalloc:#x} stream end {u.notes['compressed_end']:#x}")

File: src/data/pklite.ts
/**
 * PKLITE 1.15 (extra compression, large model) depacker for MICRO.EXE, port of tools/mm/pklite.py, which was
 * transliterated from the stub inside the executable.
 *
 * Stub layout (image offsets, image = file minus the 0x60-byte MZ header):
 *   0x000 loader, 0x034 XOR-chain decryptor (0x136 words from 0x2B2 downwards, plain = c[i] ^ c[i+1], seed 0x070C),
 *   0x058 decompressor (prefix tables at decompressor-relative 0x21D / 0x228 / 0x233 / 0x243),
 *   0x2C0 compressed bitstream, then relocations, then SS SP CS IP.
 *
 * Bitstream: 16-bit LE words, LSB first. Flag 0 = literal (byte ^ remaining-bit counter), flag 1 = match.
 * Lengths come from a prefix tree with escape 0x19 (+ raw byte; 0xFE = no-op, 0xFF = end). Offsets: high byte from
 * a prefix tree, low byte raw. Relocations: [count][offset * count] blocks, count 0 advances the segment by 0x0FFF
 * paragraphs, 0xFFFF terminates.
 */

export interface UnpackedExe {
  /** Load image (segment 0 = load segment). */
  image: Uint8Array;
  relocs: { segment: number; offset: number }[];
  ss: number; sp: number; cs: number; ip: number;
}

const HEADER_LEN = 0x60, COMPRESSED_START = 0x2C0, DECRYPT_TOP = 0x2B2, DECRYPT_WORDS = 0x136, DECRYPT_SEED = 0x070C;
const DECOMP_BASE = 0x58;

class BitReader {
  private buf = 0;
  left = 0;
  constructor(private readonly data: Uint8Array, public pos: number) { this.reload(); }
  private reload(): void { this.buf = this.data[this.pos]! | (this.data[this.pos + 1]! << 8); this.pos += 2; this.left = 16; }
  bit(): number { const b = this.buf & 1; this.buf >>= 1; if (--this.left === 0) this.reload(); return b; }
  byte(): number { return this.data[this.pos++]!; }
  word(): number { const v = this.data[this.pos]! | (this.data[this.pos + 1]! << 8); this.pos += 2; return v; }
}

function decryptStub(image: Uint8Array): void {
  let prev = DECRYPT_SEED, off = DECRYPT_TOP;
  for (let i = 0; i < DECRYPT_WORDS; i++) {
    const w = image[off]! | (image[off + 1]! << 8), p = w ^ prev;
    image[off] = p & 0xFF; image[off + 1] = p >> 8;
    prev = w; off -= 2;
  }
}

/** Returns the match length, -1 for the 0xFE no-op, or undefined at the end marker. */
function readLength(br: BitReader, t1: Uint8Array, t2: Uint8Array, t3: Uint8Array): number | undefined {
  let bx = br.bit();
  bx = (bx << 1) | br.bit();
  if (bx >= 2) return bx;
  bx = (bx << 1) | br.bit();
  if (bx === 0) return t1[0]!;
  bx = (bx << 1) | br.bit();
  if (bx < 5) return t1[bx]!;
  bx = (bx << 1) | br.bit();
  if (bx <= 0xC) return t1[bx]!;
  bx &= 3;
  bx = (bx << 1) | br.bit();
  let cl: number;
  if (bx < 5) cl = t2[bx]!;
  else {
    bx = (bx << 1) | br.bit();
    if (bx <= 0xC) cl = t2[bx]!;
    else {
      bx &= 3;
      bx = (bx << 1) | br.bit();
      if (bx < 5) cl = t3[bx]!;
      else { bx = (bx << 1) | br.bit(); cl = t3[bx]!; }
    }
  }
  if (cl !== 0x19) return cl;
  const al = br.byte();
  if (al >= 0xFE) return al === 0xFF ? undefined : -1;
  return 0x19 + al;
}

function readOffsetHigh(br: BitReader, toff: Uint8Array): number {
  if (br.bit()) return 0;
  let bx = 0;
  for (let i = 0; i < 3; i++) bx = (bx << 1) | br.bit();
  if (bx < 2) return toff[bx]!;
  bx = (bx << 1) | br.bit();
  if (bx < 8) return toff[bx]!;
  bx = (bx << 1) | br.bit();
  if (bx < 0x17) return toff[bx]!;
  bx = (bx << 1) | br.bit();
  return bx & 0xDF;
}

export function isPkliteExe(data: Uint8Array): boolean {
  if (data.length < 0x60 || data[0] !== 0x4D || data[1] !== 0x5A) return false;
  const hdr = (data[8]! | (data[9]! << 8)) * 16;
  return hdr === HEADER_LEN && data[0x1C] === 0x0F && (data[0x1D]! & 0x30) === 0x30;
}

export function unpackPklite(data: Uint8Array): UnpackedExe {
  if (!isPkliteExe(data)) throw new Error('expected a PKLITE 1.15 executable (extra compression, large model)');
  const image = data.slice(HEADER_LEN);
  decryptStub(image);
  const t1 = image.subarray(DECOMP_BASE + 0x21D, DECOMP_BASE + 0x22D), t2 = image.subarray(DECOMP_BASE + 0x228, DECOMP_BASE + 0x238);
  const t3 = image.subarray(DECOMP_BASE + 0x233, DECOMP_BASE + 0x243), toff = image.subarray(DECOMP_BASE + 0x243, DECOMP_BASE + 0x263);

  let out = new Uint8Array(0x20000), n = 0;
  const push = (b: number): void => {
    if (n === out.length) { const g = new Uint8Array(out.length * 2); g.set(out); out = g; }
    out[n++] = b;
  };
  const br = new BitReader(image, COMPRESSED_START);
  for (;;) {
    if (br.bit() === 0) { push(br.byte() ^ (br.left & 0xFF)); continue; }
    const length = readLength(br, t1, t2, t3);
    if (length === undefined) break;
    if (length === -1) continue;
    const off = length === 2 ? br.byte() : (readOffsetHigh(br, toff) << 8) | br.byte();
    if (off === 0 || off > n) throw new Error(`bad match offset ${off} at ${n}`);
    const start = n - off;
    for (let i = 0; i < length; i++) push(out[start + i]!);
  }

  const relocs: UnpackedExe['relocs'] = [];
  let seg = 0;
  for (;;) {
    const count = br.word();
    if (count === 0xFFFF) break;
    if (count === 0) { seg += 0x0FFF; continue; }
    for (let i = 0; i < count; i++) relocs.push({ segment: seg, offset: br.word() });
  }
  const ss = br.word(), sp = br.word(), cs = br.word(), ip = br.word();
  return { image: out.slice(0, n), relocs, ss, sp, cs, ip };
}
